current/runtime/storage.md

```diff
@@ -4,22 +4,128 @@ lang: en
+single primitive value, a single `struct` or a single collection of related items. Although wrapping related items in
+a shared `struct` is an excellent way to reduce the number of storage reads (an important consideration), at some point
+the size of the object will begin to incur costs that may outweigh the optimization in storage reads. Storage values
+can be used to store lists of items, but runtime developers should take care with respect to the size of these lists.
+Large lists incur storage costs just like large `structs`. Furthermore, iterating over a large list in your runtime may
+result in exceeding the block production time.
+
+#### Methods
+
+Refer to the Storage Value documentation for
+[a comprehensive list of the methods that Storage Values expose](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#required-methods).
+Some of the most important methods are summarized here:
+
+##### `get()`
+
+Load the value from storage.
+
+##### `put(val)`
+
+Store the provided value.
+
+##### `mutate(fn)`
+
+Mutate the value with the provided function.
+
+##### `take()`
+
+Remove the value from storage.
+
+### Storage Maps
+
+Storage Maps are implemented as hash maps, which is a pattern that should be familiar to most developers. In order to
+give blockchain engineers increased control over the way in which these data structures are stored, Substrate allows
+developers to select the hashing algorithm that is used to generate map keys. Map data structures are ideal for managing
+sets of items whose elements will be accessed randomly, as opposed to iterating over them sequentially in their entirety.
+
+#### Methods
+
+[Storage Maps expose an API](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#required-methods)
+that is similar to that of Storage Values. The selected methods referenced below all use a single key; use two keys for
+[Storage Double Maps](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#required-methods):
+
+##### `get(key)`
+
+Load the value associated with the provided key from storage.
+
+##### `store(key, val)`
+
+Store the provided value by associating it with the given key.
+
+##### `mutate(key, fn)`
+
+Use the provided function to mutate the value associated with the given key.
+
+##### `take(key)`
+
+Remove the value associated with the given key from storage.
+
+### Iterable Storage Maps
+
+The Substrate storage API provides iterable map implementations. Because maps are often used to track unbounded sets of
+data (account balances, for example) it is especially likely to exceed block production time by iterating over maps in
+their entirety within the runtime. Furthermore, because maps are comprised of more layers of indirection than native
+lists, they are significantly more costly than lists to iterate over with respect to time. Depending on
```

Comment:
I think we can just explain what is the "more layers of indirection than native lists", basically iterating over a map is just iterating over all keys after the prefix `storage_prefix` so this trie iteration is more costy than native list because it is a trie.

Reply:
I felt like that was a step too far for this document and that topics like that belong in [the advanced storage documentation](https://github.com/substrate-developer-hub/knowledgebase/blob/runtime-storage/current/advanced/storage.md).

Reply:
The above document just explain how the trie works not how frame_support storage abstractions are implemented using the trie, no ?
As a user in order to understand more precisely the cost of those abstraction, explaining their implementation on the trie seems useful (maybe it belong somewhere else in the doc but still I feel like saying that a storage map just store all its value after a prefix and iterating upon means iterating the trie after a prefix is quite straightforward and quickly understandable)

> Because maps are often used to track unbounded sets of
data it is especially likely to exceed block production time by iterating over maps in
their entirety within the runtime

This is just a trade off, some storage map in the runtime are quite bounded actually. But we use storage map because they are more often accessed by individual value than as a whole. Thus using a storage map is more efficient because accessing one value is cheaper, but it is sometime iterated upon still because it is not too big.

Reply:
Tried to clarify this in the docs :+1: